source/common/websocket/codec.cc

```diff
@@ -0,0 +1,220 @@
+  }
+
+  if (num_remaining_extended_length_bytes_ == 0) {
+    if (num_remaining_masking_key_bytes_ > 0) {
+      state_ = State::FrameHeaderMaskingKey;
+    } else {
+      frameDataStart();
+    }
+  }
+  return bytes_decoded;
+}
+
+uint8_t Decoder::doDecodeMaskingKey(absl::Span<const uint8_t>& data) {
+  if (!frame_.masking_key_.has_value()) {
+    frame_.masking_key_ = 0;
+  }
+  uint8_t bytes_decoded = 0;
+  uint64_t bytes_to_decode = data.length() <= num_remaining_masking_key_bytes_
+                                 ? data.length()
+                                 : num_remaining_masking_key_bytes_;
+  while (bytes_to_decode > 0) {
+    frame_.masking_key_.value() |= static_cast<uint32_t>(data.at(bytes_decoded))
+                                   << 8 * (num_remaining_masking_key_bytes_ - 1);
+    num_remaining_masking_key_bytes_--;
+    bytes_to_decode--;
+    bytes_decoded++;
+  }
+  if (num_remaining_masking_key_bytes_ == 0) {
+    frameDataStart();
+  }
+  return bytes_decoded;
+}
+
+uint64_t Decoder::doDecodePayload(absl::Span<const uint8_t>& data) {
+  uint64_t remain_in_buffer = data.length();
+  uint64_t bytes_decoded = 0;
+  if (remain_in_buffer <= length_) {
+    frameData(data.data(), remain_in_buffer);
+    bytes_decoded += remain_in_buffer;
+    length_ -= remain_in_buffer;
+  } else {
+    frameData(data.data(), length_);
+    bytes_decoded += length_;
+    length_ = 0;
+  }
+  if (length_ == 0) {
+    state_ = State::FrameFinished;
+  }
+  return bytes_decoded;
+}
+
+absl::optional<std::vector<Frame>> Decoder::decode(Buffer::Instance& input) {
+  absl::optional<std::vector<Frame>> output = absl::nullopt;
+  uint64_t bytes_consumed_by_frame = 0;
+  resetDecoder();
+  for (const Buffer::RawSlice& slice : input.getRawSlices()) {
+    absl::Span<const uint8_t> data(reinterpret_cast<uint8_t*>(slice.mem_), slice.len_);
+    while (!data.empty()) {
+      uint64_t bytes_decoded = 0;
+      switch (state_) {
+      case State::FrameHeaderFlagsAndOpcode:
+        bytes_decoded = doDecodeFlagsAndOpcode(data);
+        if (bytes_decoded == 0) {
+          return absl::nullopt;
+        }
+        break;
+      case State::FrameHeaderMaskFlagAndLength:
+        bytes_decoded = doDecodeMaskFlagAndLength(data);
+        break;
+      case State::FrameHeaderExtendedLength:
+        bytes_decoded = doDecodeExtendedLength(data);
+        break;
+      case State::FrameHeaderMaskingKey:
+        bytes_decoded = doDecodeMaskingKey(data);
+        break;
+      case State::FramePayload:
+        bytes_decoded = doDecodePayload(data);
+        break;
+      case State::FrameFinished:
+        frameDataEnd(bytes_consumed_by_frame, input, output);
+        break;
+      }
+      data.remove_prefix(bytes_decoded);
+      bytes_consumed_by_frame += bytes_decoded;
+    }
+    // Handles when slice ended with a complete frame scenario
+    if (state_ == State::FrameFinished) {
+      frameDataEnd(bytes_consumed_by_frame, input, output);
+    }
```

Comment:
Hm, it'd be nice if we could avoid this since this is basically identical to the `case State::FrameFinished:` block earlier. Can we change `while (!data.empty()) {` to while (!data.empty() || state_ != State::FrameFinished) {`? I *think* that will do what we want.